### DISCUSSION TITLE
DeferredReflection and Pickled MetaData

### DISCUSSION THREAD

**Discussion by paultatulea** [sqlalchemy/sqlalchemy#9942] [Usage Questions] [answered]
I want to use declarative mapping to define my models and set the engine for reflection after declaring the models. However, I also want to pickle the metadata after reflecting, and use that metadata instead of reflecting all the tables each time.

When you run the below script the first time, it runs with no issues. However, on the second run, the following error occurs:

`sqlalchemy.orm.exc.UnmappedClassError: Class __main__.Foo is a subclass of DeferredReflection.  Mappings are not produced until the .prepare() method is called on the class hierarchy.`

How can I define my models such that I don't need to first read the metadata or create the engine before the models are imported?

```python
import os
import pickle
import sqlalchemy
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.declarative import DeferredReflection
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm import configure_mappers


Base = declarative_base(cls=DeferredReflection)


class Foo(Base):
    __tablename__ = "foo"


def main():
    engine = sqlalchemy.create_engine("sqlite:///test.db")

    with engine.connect() as conn:
        conn.execute(
            sqlalchemy.text(
                "create table if not exists foo ( "
                "id int primary key, "
                "description varchar(255) "
                "); "
            )
        )

    if os.path.exists("metadata.pkl"):
        with open("metadata.pkl", "rb") as f:
            metadata = pickle.load(f)
            Base.metadata = metadata
        configure_mappers()
    else:
        Base.prepare(engine)
        with open("metadata.pkl", "wb") as f:
            pickle.dump(Base.metadata, f)

    Session = sessionmaker(bind=engine)

    with Session() as session:
        results = session.execute(sqlalchemy.select(Foo)).scalars().fetchall()
        print("Query results:", results)


main()
```

I tried the following with no success:

```python
Base = declarative_base()

class Foo(DeferredReflection, Base):
    __tablename__ = "foo"

def main():
    ...
    if os.path.exists("metadata.pkl"):
        for klass in DeferredReflection.__subclasses__():
            new_bases = tuple(
                base for base in klass.__bases__ if base is not DeferredReflection
            )
            new_class = type(klass.__name__, new_bases, dict(klass.__dict__))
            globals()[klass.__name__] = new_class
        with open("metadata.pkl", "rb") as f:
            metadata = pickle.load(f)
            Base.metadata = metadata
        configure_mappers()
    else:
        DeferredReflection.prepare(engine)
        with open("metadata.pkl", "wb") as f:
            pickle.dump(Base.metadata, f)
```

```
path\to\pkg\a.py:35: SAWarning: This declarative base already contains a class with the same class name and module name as __main__.Foo, and will be replaced in the string-lookup table.
  new_class = type(klass.__name__, new_bases, dict(klass.__dict__))
Traceback (most recent call last):
  File "path\to\pkg\a.py", line 53, in <module>
    main()
  File "path\to\pkg\a.py", line 35, in main
    new_class = type(klass.__name__, new_bases, dict(klass.__dict__))
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\decl_api.py", line 76, in __init__
    _as_declarative(reg, cls, dict_)
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\decl_base.py", line 126, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\decl_base.py", line 183, in setup_mapping
    return cfg_cls(registry, cls_, dict_, table, mapper_kw)
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\decl_base.py", line 335, in __init__
    self._early_mapping(mapper_kw)
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\decl_base.py", line 215, in _early_mapping
    self.map(mapper_kw)
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\decl_base.py", line 1047, in map
    mapper_cls(self.cls, self.local_table, **self.mapper_args),
  File "<string>", line 2, in __init__
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\util\deprecations.py", line 375, in warned
    return fn(*args, **kwargs)
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\mapper.py", line 690, in __init__
    self._configure_class_instrumentation()
  File "path\to\pkg\venv\lib\site-packages\sqlalchemy\orm\mapper.py", line 1277, in _configure_class_instrumentation
    assert manager.class_ is self.class_
AssertionError
```

**Comment by zzzeek** [comment]
the moment you created the Foo class, it made its `Table` object inside of `Base.metadata`, so when you replaced that later, the `Table` to which `Foo` is mapped is empty.    The way deferredreflection works is it reflects into the existing `Table` objects using [extend_existing](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.params.extend_existing).

There's not a spectacular way to fill in the pickled tables here unless you used [Table.to_metadata](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.to_metadata). which would work here.  But pickling metadata is awkward and to_metadata might have some awkwardness also.

What's the actual reason you can't run reflection each time?    Reflection is quite fast in 2.0.

**Reply by paultatulea** [reply]
Hi Mike, thank you for the quick response! It is a legacy code base that is being upgraded so has only recently been updated to 1.4 to bridge the gap. I think that performance was the only real concern. In 2.0, is there no performance advantage to pickling and loading over reflecting each time, even in a rather large database with hundreds of tables?

I haven't looked into Table.to_metadata, thanks for pointing this out. In my use case, I would load the pickled MetaData object, iterate over all the tables in the metadata, and call `to_metadata` on each table with the target being the new MetaData object?

**Reply by CaselIT** [reply]
> Would you say that in 2.0, there is no performance advantage to pickling and loading over reflecting each time, even in a rather large database with hundreds of tables?

reflection on v2 has become faster, especially on postgress and oracle, but it's probably still slower than loading from file. For a db that size it should probably be less than 10s

**Reply by paultatulea** [reply]
Okay that's very good to know, will look into the possibility of upgrading as soon as possible and bench marking my loading from file vs reflecting. With regard to `Table.to_metadata`, I'm not quite sure how to use this so that the `DeferredReflection` base class does not raise the `UnmappedClassError` when I try to use the model.

Do I set the `__table__` attribute of the model to the result of `Table.to_metadata(new_metadata)` for each table in the loaded MetaData?

**Reply by zzzeek** [reply, marked as answer]
> Hi Mike, thank you for the quick response! It is a legacy code base that is being upgraded so has only recently been updated to 1.4 to bridge the gap. I think that performance was the only real concern. In 2.0, is there no performance advantage to pickling and loading over reflecting each time, even in a rather large database with hundreds of tables?

haven't done benching but pickling is likely to be "faster" in most/all cases but still not very fast.   if this is like startup for a long running server, startup between 1 second or 5 seconds might be not much of a difference.     


> I haven't looked into Table.to_metadata, thanks for pointing this out. In my use case, I would load the pickled MetaData object, iterate over all the tables in the metadata, and call `to_metadata` on each table with the target being the new MetaData object?

that would be the idea but I just tested and it looks like to_metadata does not actually merge the tables right now, it skips the ones that exist, so it cannot be used in the way I suggested.  nobody has ever asked for this use case before.  if we really wanted to support this I'd likely have it pickle the `Inspector` object, or part of its state, instead since it would make for a much simpler pickle.      Looking at the source right now there's no path to feed that in without changing a lot of code.

Another very exotic way this could be made to work would be temporary, global monkeypatching of the `Inspector` object to memoize all the method calls and return values, then patch again on top of that with pickled results so that no database calls occur when the patched version is in place.    

other than that you need to load up this pickled file before you make the Foo class, right up when you first make the declarative Base.   that's likely the easiest path here.

**Reply by CaselIT** [reply]
regarding patching the inspector, v2 should be a lot easier to do since internally it uses the `_ReflectionInfo` to hold all the reflection data.

**Reply by paultatulea** [reply]
Thanks for digging in and testing out possible solutions!
Unfortunately not a long running application so the startup time would be a factor, however a few seconds would not make any difference so migrating to 2.0 would be best in the long run, for all the other great features and improvements as well.

I think that no one has requested it before is an indicator that something may be wrong with my approach! The goal would be to be able to define the models such that they can be imported without the need to handle the setup beforehand, and the setup can happen lazily at any time it is needed.

Loading the metadata before declaring the `Base` would be best but there is also some logic to refresh the metadata if there has been a schema update since the last pickle, in which case a reflection is needed. That requires connecting to the database to query the last schema update, so the setup needs to happen before the import there as well.

Thanks for all the help!

**Comment by CaselIT** [comment]
You may look into https://github.com/agronholm/sqlacodegen that creates the python files given a database schema
